retrospectives/2019-11-20-worker-manager-overprovisioning.md

```diff
@@ -0,0 +1,79 @@
+
+## Timeline
+  - 2019-11-20 18:06 UTC ::::: Tasks are submitted as part of a try push
+  - 2019-11-20 19:09 UTC ::::: The worker-pool is configured to use spot instances, we see this error: "Error calling AWS API: There is no Spot capacity available that matches your request."
+  - 2019-11-20, 19:38 UTC ::::: More pending tasks as part of the try push => more requestedCapacity
+  - 2019-11-20, 19:48 UTC ::::: A change is landed to allow the worker-pool to use on-demand instances
+  - 2019-11-20 19:39 UTC ::::: Another “Error calling AWS API: There is no Spot capacity available that matches your request.” (the last time we see this, indicating that the change to on-demand has happened)
+  - 2019-11-20 19:40 UTC ::::: We spawn 12 instances (each instance has capacity 4)
+  - 2019-11-20 20:01:15 UTC ::::: We’re at runningCapacity 80
+  - 2019-11-20 20:02:32 UTC ::::: We’re back down to 0 runningCapacity
+  - 2019-11-20 20:05 UTC ::::: "Error calling AWS API: We currently do not have sufficient c5.metal capacity in the Availability Zone you requested (us-west-2a). Our system will be working on provisioning additional capacity. You can currently get c5.metal capacity by not specifying an Availability Zone in your request or choosing us-west-2b, us-west-2c, us-west-2d." This error continues for quite some time, repeatedly
+
+## Fallout
+
+* Some large amount of $$$
+* Loss of trust from other teams and management
+* Lots of scramble and stress on our part in the “relaxing” period after our cloudops migration
+
+## Action Items
+
+* [DONE] The worker-pool was disabled and all workers were terminated as soon as the issue was observed.
+* [DONE] Additional logging and error reporting has been added to worker-manager to call attention to overprovisioning.
+  You can use https://gist.github.com/imbstack/b18a697fb3046a90080499a93b0ffe15 to check for overprovisioned worker-pools
+* [DONE] The configuration of worker-runner on generated AMIs has been corrected to use aws-provider.
+* [DONE] Worker-manager should track instances that it has spawned that have not registered and terminate them, as this is indicative of a configuration issue.
+* Additionally, if a worker-pool overwhelmingly spawns failing workers it should be disabled.
+* [bug 1604925] Worker-manager’s running capacity per worker-pool should be made visible in the UI
+* AMIs will be generated for the taskcluster-staging account and tested in development clusters before production and FirefoxCI use
+* Adding billing alerts
+* Accounting for orphaned instances
+* [DONE] Worker-manager’s loops (per libiterate) should ensure that only a single scanning loop is ever running at once
+* Workers view
+* All workers need to call registerWorker (worker-runner)
+  Owlish’s requirement gathering document to aid the design  https://docs.google.com/document/d/1BFha7fDTX3uXYdVbmeHCXQK6G6o8MghHmldB7Xodz0U/edit?usp=sharing 
+* [bug 1604926] Check for orphaned instances
```

Comment:
* Bug 1587511 - worker-manager: hard kill instances after 96 hours
* Bug 1587516 - worker-manager: kill idle instances and disable bad worker pool configs